linking/cid/cidLink.go

```diff
@@ -30,20 +30,30 @@ func (lnk Link) Load(ctx context.Context, lnkCtx ipld.LinkContext, na ipld.NodeA
 	if !exists {
 		return fmt.Errorf("no decoder registered for multicodec %d", lnk.Prefix().Codec)
 	}
-	var hasher bytes.Buffer // multihash only exports bulk use, which is... really inefficient and should be fixed.
-	decodeErr := mcDecoder(na, io.TeeReader(r, &hasher))
-	// Error checking order here is tricky.
-	//  If decoding errored out, we should still run the reader to the end, to check the hash.
-	//  (We still don't implement this by running the hash to the end first, because that would increase the high-water memory requirement.)
-	//   ((Which we experience right now anyway because multihash's interface is silly, but we're acting as if that's fixed or will be soon.))
-	//  If the hash is rejected, we should return that error (and even if there was a decodeErr, it becomes irrelevant).
-	if decodeErr != nil {
-		_, err := io.Copy(&hasher, r)
-		if err != nil {
-			return err
+	var hasherBytes []byte
+	var decodeErr error
+	byteBuf, ok := r.(*bytes.Buffer)
```

Comment:
I feel a little dissatisfied making a concrete type assertion like this... but do you suppose we could accomplish the same thing with a small feature detection branch using a tiny interface?

Something like:

```
type Byter interface {
    Bytes() []byte
}
if byter, ok := r.(Byter) { // ...
```

This should match `bytes.Buffer` and do the right thing, I think, and probably costs 10ms more to check but that shouldn't bother :)

Reply:
I agree with this; the extra cost should be negligible, compared to the whole allocate + copy branch.